config/database.go

```diff
@@ -0,0 +1,322 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```

Comment:
Could we move it to an specific module? I can imagine this like `config/[some-go-files]` and `config/{database,redis,etcd,dynamodb}Backend/`.

Reply:
Strictly speaking, there's nothing stopping us from doing this, but I'd love to understand more about your thinking here -- what's the proposed benefit?

Reply:
the main benefit is to have clear separation of concerns between the modules, basically i don't want to start building small `app` style directories, where you have a 35000 LOC of interrelated and interdependent code. If each backend is independent and has his own module, each backend can grow to various files without problem, and still can be pretty organized. If you have everything in the same module is very easy to start to generate interdependencies and then, the split is very difficult.

Reply:
Definitely agree that different concerns probably belong in different packages.

In this case, it seems to me that the concerns are are (intentionally) coupled and significantly overlapping, just that the backing implementations vary quite a bit. As a concrete example, were I to push out the file store and database store, I'd either need to copy various helper methods into each package, or export them on the `config` package to enable reuse, cluttering its surface API. In fact, there's already duplicate code in the two existing stores that I'd like to refactor more to ensure we don't end up with diverging semantics between the stores in the future.

In general, too, I'm no longer a fan of using go packages /just/ to "organize" code. I used to try to force some of the design patterns I had adopted in other languages into go:
```sh
somepackage/
    interface.go
    common.go
    someimplementation1/
    someimplementation2/
    mockimplementation/
```

but the big distinction in my mind with go is that there is no concept of "subpackages". `someimplementation1` isn't special just because it's a child of `somepackage` -- it could be anywhere in the $GOROOT and have the same semantics. (*)

Looking at the standard library, it seems to be idiomatic to keep the package structure relatively flat, even if that means lots of files.

Thoughts?

 (* Technically this is not true if you bring `internal` into the mix. But I'm also inclined to use `internal` at a "go module" level and not as a way to splice up packages like this.)

Reply:
Agree with @jespino here. It's a common pattern to have interface implementations as submodules. Eg: https://golang.org/pkg/hash/

Reply:
Pedantically, there is no such thing as "submodules" (or "subpackages") -- but I understand your point.

Do you have any suggestions on how to accomplish this and address the issue of sharing of code? (e.g. the internal methods I refer to above that don't belong as exports of `config` but would need to be shared by both backing implementations.) Should I go down the `internal` path?

Reply:
@jespino and I discussed offline:
* This package currently consists of eight non-test, non-data files
* Five of those files are "common" and used by file.go, database.go and memory.go
* Breaking this up into more packages seems premature at this point, given that they each would contain one file (plus tests)
* Breaking them up would also require exporting all of the common code that is currently hidden.

We both agreed that if any of these backing implementations grows out of hand, we could revisit this approach in the future, but that right now it seems to be premature to reorganize the code.